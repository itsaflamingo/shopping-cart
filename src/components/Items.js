import React, {useState, useEffect} from 'react';
import Cart from './Cart'

function Items(props) {

  const [closeCart, setCloseCart] = useState(true);
  const [cart, setCart] = useState([]);
  const [quantity, setQuantity] = useState(1);
  const [item, setItem] = useState();

  useEffect(() => {
    setQuantity(1);
  }, [item])

  const addToCart = (e, item) => setCart(cart.concat(item)) 

  const updateQuantity = () => {
    const increaseQuantity = (e, item) => {
      setItem(item);
      setQuantity(quantity+1);
      // problem. If it is included, have to click 2x for new num. 
      updateObj(e, item);
      console.log(quantity)
    };
    
    const decreaseQuantity = (e, item, num) => {
      if(item.quantity === 0) return;
      setQuantity(item.quantity);
      setQuantity(num-1);
      updateObj(e, item);
    };
    
    return {
      increaseQuantity,
      decreaseQuantity,
    }
  }

  const updateObj = (e, item) => {
    if(e === undefined) return;
      item.quantity = quantity;
  }

  return (
    <div className="items">
            {props.items.map((item) => {
                return (
                    <div className='item' key={item.id} 
                    onClick={() => setCloseCart(false)}>
                        <p>{item.name}</p>
                        <button onClick={e => addToCart(e, item)}>Add to Cart</button>
                    </div>
                )
            })}
            {!closeCart && <Cart items={cart} setCloseCart={setCloseCart} updateQuantity={updateQuantity} quantity={quantity} updateObj={updateObj} />}
    </div>
  );
}

export default Items;
