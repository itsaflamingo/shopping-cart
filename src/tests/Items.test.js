import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";  // optional
import {BrowserRouter as Router} from 'react-router-dom';
import { act } from "react-dom/test-utils";
import Items from '../components/Items'

describe('Shop items', () => {

    const items = [{
        title:'name1',
        id:'1'
    },
    {
        title: 'name2',
        id: '2'}]

    it('Renders items on page', () => {
        const {container} = render(
            <Router>
                <Items items={items} />
            </Router>) 
            
            expect(container).toMatchSnapshot();
    })
    it('Opens cart when "add to cart" button is clicked', () => {
        const {asFragment} = render(
            <Router>
                <Items items={items} />
            </Router>) 
        const firstRender = asFragment();
        const button = screen.getAllByRole('button', {name: 'Add to Cart'});
        act(() => {
            userEvent.click(button[0]);
        })
        expect(firstRender).toMatchSnapshot();
    })
    it('when item added to cart, cart info is updated', () => {
        const {container} = render(
            <Router>
                <Items items={items} />
            </Router>) 

        const button = screen.getAllByRole('button', {name: 'Add to Cart'});
        act(() => {
            userEvent.click(button[0]);
        })
        const name = screen.getAllByText('name1');
        expect(name[1].innerHTML).toBe('name1');
    })
    test.todo('when item quantity is changed, quantity displayed changes to correct amount')

})