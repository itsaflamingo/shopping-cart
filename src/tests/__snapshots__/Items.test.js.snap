// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shop items Renders items on page 1`] = `
<div>
  <div
    id="items"
  >
    <div
      class="item-container"
    >
      <div
        aria-label="item"
        class="item"
      >
        <div
          class="image"
          style="background-image: url(undefined); background-size: contain; background-repeat: no-repeat;"
        >
          <button
            id="add-to-cart"
          >
            Add to Cart
          </button>
        </div>
        <div
          class="item-info"
        >
          <p
            class="title"
          >
            name1
          </p>
          <p>
            $
          </p>
        </div>
      </div>
    </div>
    <div
      class="item-container"
    >
      <div
        aria-label="item"
        class="item"
      >
        <div
          class="image"
          style="background-image: url(undefined); background-size: contain; background-repeat: no-repeat;"
        >
          <button
            id="add-to-cart"
          >
            Add to Cart
          </button>
        </div>
        <div
          class="item-info"
        >
          <p
            class="title"
          >
            name2
          </p>
          <p>
            $
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
